isvcs/zk.go

```diff
@@ -79,39 +80,73 @@ func initZK() {
 // a health check for zookeeper
 func zkHealthCheck(halt <-chan struct{}) error {
 	for {
-		// establish a zookeeper connection
-		conn, ec, err := zk.Connect([]string{"127.0.0.1:2181"}, time.Second*10)
-		defer func() {
-			if conn != nil {
-				conn.Close()
+		select {
+		case <-halt:
+			glog.V(1).Infof("Quit healthcheck for zookeeper")
+			return nil
+		default:
+			// Try ruok.
+			ruok, err := zkFourLetterWord("127.0.0.1:2181", "ruok", time.Second*10)
+			if err != nil {
+				glog.Warningf("No response to ruok from ZooKeeper: %s", err)
+				time.Sleep(1 * time.Second)
+				continue
 			}
-		}()
-		if err != nil {
-			glog.V(1).Infof("Could not connect to zookeeper: %s", err)
-			time.Sleep(1 * time.Second)
-		} else {
-			//wait for session
-			sesstionTimeout := 5 * time.Second
-			sessionTimer := time.NewTimer(sesstionTimeout)
-			defer sessionTimer.Stop()
-			timedOut := false
-			for !timedOut {
-				select {
-				case e := <-ec:
-					if e.State == zk.StateHasSession {
-						// success
-						glog.V(1).Infof("Zookeeper running, browser at http://localhost:12181/exhibitor/v1/ui/index.html")
-						return nil
-					}
-				case <-halt:
-					glog.V(1).Infof("Quit healthcheck for zookeeper")
-					return nil
-				case <-sessionTimer.C:
-					//Fall through loop and try again
-					glog.V(1).Infof("ZK Session not available in %s", sesstionTimeout)
-					timedOut = true
-				}
+
+			// ruok should respond either with "imok" or not at all.
+			// If for some reason that isn't the case, there's a problem.
+			glog.V(2).Infof("ruok: \"%s\"", ruok)
+			if string(ruok) != "imok" {
+				glog.Warningf("Improper response to ruok from ZooKeeper: %s", ruok)
+				time.Sleep(1 * time.Second)
+				continue
+			}
+
+			// Since ruok works, try stat next.
+			stat, err := zkFourLetterWord("127.0.0.1:2181", "stat", time.Second*10)
+			if err != nil {
+				glog.Warningf("No response to stat from ZooKeeper: %s", err)
+				time.Sleep(1 * time.Second)
+				continue
 			}
+
+			// If we get "This ZooKeeper instance is not currently serving requests", we know it's waiting for quorum and can at least note that in the logs.
+			glog.V(2).Infof("stat: \"%s\"", stat)
+			if string(stat) == "This ZooKeeper instance is not currently serving requests\n" {
+				glog.Warningf("ZooKeeper is running, but still establishing quorum.")
+				time.Sleep(1 * time.Second)
```

Comment:
We have time.sleep 1 sec for all continuation, should this be moved to outer loop (e.g. first or last statement of the for loop block)?